SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp

```diff
@@ -0,0 +1,308 @@
+	}
+#endif // TRACE_LIB_ACTIVE
+	return false;
+}
+
+bool USpatialLatencyTracer::EndLatencyTrace(UObject* WorldContextObject)
+{
+#if TRACE_LIB_ACTIVE
+	if (USpatialLatencyTracer* Tracer = GetTracer(WorldContextObject))
+	{
+		return Tracer->EndLatencyTrace_Internal();
+	}
+#endif // TRACE_LIB_ACTIVE
+	return false;
+}
+
+#if TRACE_LIB_ACTIVE
+bool USpatialLatencyTracer::IsValidKey(const TraceKey& Key)
+{
+	FScopeLock Lock(&Mutex);
+	return TraceMap.Find(Key);
+}
+
+TraceKey USpatialLatencyTracer::GetTraceKey(const UObject* Obj, const UFunction* Function)
+{
+	FScopeLock Lock(&Mutex);
+
+	ActorFuncKey FuncKey{ Cast<AActor>(Obj), Function };
+	TraceKey ReturnKey = InvalidTraceKey;
+	TrackingTraces.RemoveAndCopyValue(FuncKey, ReturnKey);
+	return ReturnKey;
+}
+
+void USpatialLatencyTracer::WriteToLatencyTrace(const TraceKey& Key, const FString& TraceDesc)
+{
+	FScopeLock Lock(&Mutex);
+
+	if (TraceSpan* Trace = TraceMap.Find(Key))
+	{
+		WriteKeyFrameToTrace(Trace, TraceDesc);
+	}
+}
+
+void USpatialLatencyTracer::EndLatencyTrace(const TraceKey& Key, const FString& TraceDesc)
+{
+	FScopeLock Lock(&Mutex);
+
+	if (TraceSpan* Trace = TraceMap.Find(Key))
+	{
+		WriteKeyFrameToTrace(Trace, TraceDesc);
+
+		Trace->End();
+		TraceMap.Remove(Key);
+	}
+}
+
+void USpatialLatencyTracer::WriteTraceToSchemaObject(const TraceKey& Key, Schema_Object* Obj)
+{
+	FScopeLock Lock(&Mutex);
+
+	if (TraceSpan* Trace = TraceMap.Find(Key))
+	{
+		const improbable::trace::SpanContext& TraceContext = Trace->context();
+		improbable::trace::TraceId _TraceId = TraceContext.trace_id();
+		improbable::trace::SpanId _SpanId = TraceContext.span_id();
+
+		SpatialGDK::AddBytesToSchema(Obj, SpatialConstants::UNREAL_RPC_TRACE_ID, &_TraceId[0], _TraceId.size());
+		SpatialGDK::AddBytesToSchema(Obj, SpatialConstants::UNREAL_RPC_SPAN_ID, &_SpanId[0], _SpanId.size());
+	}
+}
+
+TraceKey USpatialLatencyTracer::ReadTraceFromSchemaObject(Schema_Object* Obj)
+{
+	FScopeLock Lock(&Mutex);
+
+	check(GetActiveTrace() == nullptr);
+
+	const uint8* TraceBytes = Schema_GetBytes(Obj, SpatialConstants::UNREAL_RPC_TRACE_ID);
+	const uint8* SpanBytes = Schema_GetBytes(Obj, SpatialConstants::UNREAL_RPC_SPAN_ID);
+
+	improbable::trace::TraceId _TraceId;
+	memcpy(&_TraceId[0], TraceBytes, sizeof(improbable::trace::TraceId));
+
+	improbable::trace::SpanId _SpanId;
+	memcpy(&_SpanId[0], SpanBytes, sizeof(improbable::trace::SpanId));
+
+	improbable::trace::SpanContext DestContext(_TraceId, _SpanId);
+
+	TraceSpan RetrieveTrace = improbable::trace::Span::StartSpanWithRemoteParent("Read Trace From Schema Obj", DestContext);
+	TraceMap.Add(ActiveTraceKey, MoveTemp(RetrieveTrace));
+
+	return ActiveTraceKey;
+}
+
+USpatialLatencyTracer* USpatialLatencyTracer::GetTracer(UObject* WorldContextObject)
+{
+	UWorld* World = GEngine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::ReturnNull);
+	if (World == nullptr)
```

Comment:
Why do we sometimes use `WorldContextObject` to get world, and sometimes use `GWorld`? Is it potentially buggy?

Reply:
Yes - I'm looking into this currently. It shouldn't be at runtime, however within the editor there's a change the World will be incorrect. Watch this space.

Reply:
So GWorld will always be valid when running standalone, and in editor will be updated to point to the world currently being ticked inside `UGameEngine::Tick`. This path is currently invoked by RPCPayload (as we have no handle on any Unreal object when it's deserialized).